cmd/containerd-shim-runc-v2/task/service.go

```diff
@@ -677,28 +634,25 @@ func (s *service) processExits() {
 		// Handle the exit for a created/started process. If there's more than
 		// one, assume they've all exited. One of them will be the correct
 		// process.
-		var cps, skipped []containerProcess
-		for _, cp := range s.running[e.Pid] {
-			_, init := cp.Process.(*process.Init)
-			if init && s.pendingExecs[cp.Container] != 0 {
-				// This exit relates to a container for which we have pending execs. In
-				// order to ensure order between execs and the init process for a given
-				// container, skip processing this exit here and let the `handleStarted`
-				// closure for the pending exec publish it.
-				skipped = append(skipped, cp)
-			} else {
-				cps = append(cps, cp)
-			}
-		}
-		if len(skipped) > 0 {
-			s.running[e.Pid] = skipped
-		} else {
-			delete(s.running, e.Pid)
-		}
+		cps := s.running[e.Pid]
+		delete(s.running, e.Pid)
 		s.lifecycleMu.Unlock()
 
 		for _, cp := range cps {
-			s.handleProcessExit(e, cp.Container, cp.Process)
+			cp := cp
+			e := e
+			_, init := cp.Process.(*process.Init)
+			if init {
+				go func() {
+					for s.pendingExecs[cp.Container] != 0 {
+						time.Sleep(time.Millisecond * 50)
+					}
+					s.handleProcessExit(e, cp.Container, cp.Process)
+				}()
+			} else {
+				s.pendingExecs[cp.Container]--
```

Comment:
Why are we decreasing `s.pendingExecs` here as well? `s.pendingExecs` keeps track of **pending** execs, not **running** execs.

If we're changing `s.pendingExecs` to account for **running** execs, we should:
1. rename `pendingExecs`, and
2. discuss whether we want to impose the additional ordering between execs and init exits even in the shared-pid namespace scenario.